crates/bevy_ecs/src/schedule/executor_parallel.rs

```diff
@@ -166,70 +166,124 @@ impl ParallelExecutor {
+
+                let mut run = move || {
                     #[cfg(feature = "trace")]
-                    let system_guard = system_span.enter();
+                    let _system_guard = system_span.enter();
                     unsafe { system.run_unsafe((), world) };
-                    #[cfg(feature = "trace")]
-                    drop(system_guard);
-                    finish_sender
-                        .send(index)
-                        .await
-                        .unwrap_or_else(|error| unreachable!("{}", error));
                 };
 
-                #[cfg(feature = "trace")]
-                let task = task.instrument(overhead_span);
-                if system_data.is_send {
-                    scope.spawn(task);
+                if can_start {
+                    let task = async move {
+                        run();
+                        finish_sender
+                            .send(index)
+                            .await
+                            .unwrap_or_else(|error| unreachable!("{}", error));
+                    };
+
+                    #[cfg(feature = "trace")]
+                    let task = task.instrument(overhead_span);
+                    if system_data.is_send {
+                        scope.spawn(task);
+                    } else {
+                        scope.spawn_local(task);
+                    }
+
+                    #[cfg(test)]
+                    {
+                        started_systems += 1;
+                    }
+
+                    self.running.insert(index);
+                    if !system_data.is_send {
+                        self.non_send_running = true;
+                    }
+                    // Add this system's access information to the active access information.
+                    self.active_archetype_component_access
+                        .extend(&system_data.archetype_component_access);
                 } else {
-                    scope.spawn_local(task);
+                    let start_listener = system_data.start.listen();
+                    let task = async move {
+                        start_listener.await;
+                        run();
+                        finish_sender
+                            .send(index)
+                            .await
+                            .unwrap_or_else(|error| unreachable!("{}", error));
+                    };
+
+                    #[cfg(feature = "trace")]
+                    let task = task.instrument(overhead_span);
+                    if system_data.is_send {
+                        scope.spawn(task);
+                    } else {
+                        scope.spawn_local(task);
+                    }
                 }
             }
             // Queue the system if it has no dependencies, otherwise reset its dependency counter.
             if system_data.dependencies_total == 0 {
-                self.queued.insert(index);
+                if !can_start {
```

Comment:
I don't like how far this logic is from effectively the "else" of this if statement. Systems that can start were already started above, but it's hard to piece that together.

Reply:
Moved it up.